Improve apiserver SLI metric name

dgrisonnet opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Add new kube-apiserver SLI metric better reflecting that the metric is
an SLI and not an SLO and deprecate the existing
apiserver_request_slo_duration_seconds in 1.27. Although the metric is
still in alpha, we prefer deprecating it for one release since it is a
critical metric used for SLOs and to make sure that users that are using
it have time to make the transition.
    
Going forward we prefer going with SLI-specific metrics, we will use
_sli_ instead of _slo_ so for consistency purposes.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Deprecate the apiserver_request_slo_duration_seconds metric for v1.27 in favor of apiserver_request_sli_duration_seconds for naming consistency purposes with other SLI-specific metrics and to avoid any confusion between SLOs and SLIs.
```

dgrisonnet committed:
metrics: improve apiserver SLI metric name

Add new kube-apiserver SLI metric better reflecting that the metric is
an SLI and not an SLO and deprecate the existing
apiserver_request_slo_duration_seconds in 1.27. Although the metric is
still in alpha, we prefer deprecating it for one release since it is a
critical metric used for SLOs and to make sure that users that are using
it have time to make the transition.

Going forward we prefer going with SLI specific metrics, we will use
_sli_ instead of _slo_ so for consistency purposes.

Signed-off-by: Damien Grisonnet <dgrisonn@redhat.com>

dgrisonnet commented:
/assign @logicalhan

dgrisonnet commented:
/sig instrumentation

logicalhan commented:
Even though this is alpha, I'm hesitant to rename it because I actually suspect people have already started using it.

dgrisonnet commented:
Yes, there should already be quite a huge amount of users since the whole ecosystem that imports https://github.com/kubernetes-monitoring/kubernetes-mixin is already using this metric.

My main reason for changing the name is that metrics are hard to discover by default since we don't have a solid documentation upstream. So the principal way to discover metrics is to look for certain patterns in the query platform. Let's say a user is looking for SLI metrics, then by searching for the `sli` pattern if our metrics have consistent names then they should get all our SLI-related metrics.

As for actually breaking the users, I think it should be fine since the metric is still ALPHA and I labeled this PR as an API change so it will appear as a breaking-change in the changelog.

fedebongio commented:
/triage accepted

logicalhan commented:
> As for actually breaking the users, I think it should be fine since the metric is still ALPHA and I labeled this PR as an API change so it will appear as a breaking-change in the changelog.

Absolutely not, I am not in favor of just breaking users. Instead, we should introduce a second metric and deprecate this one and give people an actual chance to migrate.

dgrisonnet commented:
/kind cleanup

logicalhan commented:
/assign @wojtek-t

logicalhan commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -112,6 +112,21 @@ var (
+			Buckets: []float64{0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3,
+				4, 5, 6, 8, 10, 15, 20, 30, 45, 60},
+			StabilityLevel:    compbasemetrics.ALPHA,
+			DeprecatedVersion: "1.26.0",
```
this should be 1.27

dgrisonnet replied:
surely :man_facepalming:

dgrisonnet commented:
@logicalhan makes sense, I addressed your comment.

@wojtek-t can you please also take a look?

My main concern now that we have added a deprecation period is that we will have to live with 3 SLO metrics in 1.26 which will increase the number of timeseries exposed by Kubernetes by quite a lot since these metrics are already the most expensive one. We can expect to have users complaining about the increase, but since it is only for one release, I guess it should be fine.

logicalhan commented:
lgtm but i also share cardinality concerns. 

@wojtek-t what do you think?

wojtek-t commented:
> lgtm but i also share cardinality concerns.
> @wojtek-t what do you think?

Just to double check I didn't miss anything - it's effectively the exact same metric - just we're naming it differently?
Assuming I'm right, I agree with the cardinality concern, but I don't see a better way of doing that (if we really want to rename it which makes sense to me) - just switching the name sounds too problematic to me.

So I'm fine with that, but I would like to get an ACK from at least one of folks who use that extensively internally:
@mborsz @marseel

marseel commented:
/lgtm

logicalhan reviewed (approved):
/lgtm
/approve
